### PR TITLE
support for `source_url_pattern` in config being a function

### PR DESCRIPTION
**Pull request by mayel**
👋 Having a use case of generating one set of docs for a project + some of its dependencies, this allows providing a function instead of a single pattern, so we can link to the source within each dep's repo.

Always open to suggestions about how else I could go about this...

### CODE REVIEW

File: lib/ex_doc/retriever.ex [hunk 1]
```diff
@@ -330,8 +330,19 @@ defmodule ExDoc.Retriever do
   defp source_link(%{path: _, url: nil}, _line), do: nil
 
   defp source_link(source, line) do
-    source_url = Regex.replace(~r/%{path}/, source.url, source.path)
-    Regex.replace(~r/%{line}/, source_url, to_string(line))
+    source_url_pattern(source.url, source.path, to_string(line))
+  end
+
+  def source_url_pattern(source_url_pattern, path, line) do
```
Comment (josevalim) [inline lib/ex_doc/retriever.ex:336]: Can you please move this to `ExDoc.Utils`? Add a basic unit test and we are good to go!

**Comment by mayel** [comment]
@josevalim done :)

**Comment by josevalim** [comment]
Please run `mix format` so CI is happy? :)

**Comment by mayel** [comment]
CI now passing

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**